packages/medusa-plugin-permissions/src/services/__tests__/permission.js

```diff
@@ -0,0 +1,330 @@
+      )
+      expect(accessDenied).toEqual(false)
+    })
+  })
+
+  describe("retrieveRole", () => {
+    let result
+    const permissionService = new PermissionService({
+      roleModel: RoleModelMock,
+    })
+    beforeAll(async () => {
+      jest.clearAllMocks()
+
+      result = await permissionService.retrieveRole("product_editor")
+    })
+
+    it("calls permission model functions", () => {
+      expect(RoleModelMock.findOne).toHaveBeenCalledTimes(1)
+    })
+
+    it("successfully fetches product editor permissions", () => {
+      expect(result).toEqual(permissions.productEditorPermission)
+    })
+
+    it("throws if role with name does not exist", async () => {
+      try {
+        await permissionService.retrieveRole("product_editor")
+      } catch (error) {
+        expect(error.message).toEqual(
+          "test_editor does not exist. Use method createRole to create it."
+        )
+      }
+    })
+  })
+
+  describe("createRole", () => {
+    const permissionService = new PermissionService({
+      roleModel: RoleModelMock,
+    })
+
+    beforeAll(async () => {
+      jest.clearAllMocks()
+
+      const contentEditorPermissions = [
+        {
+          method: "POST",
+          endpoint: "/contents",
+        },
+        {
+          method: "GET",
+          endpoint: "/contents",
+        },
+        {
+          method: "PUT",
+          endpoint: "/contents",
+        },
+      ]
+
+      await permissionService.createRole(
+        "content_editor",
+        contentEditorPermissions
+      )
+    })
+
+    it("calls permission model functions", () => {
+      expect(RoleModelMock.create).toHaveBeenCalledTimes(1)
+      expect(RoleModelMock.create).toHaveBeenCalledWith({
+        name: "content_editor",
+        permissions: [
+          {
+            method: "POST",
+            endpoint: "/contents",
+          },
+          {
+            method: "GET",
+            endpoint: "/contents",
+          },
+          {
+            method: "PUT",
+            endpoint: "/contents",
+          },
+        ],
+      })
+    })
+
+    it("throws if any permission is invalid", async () => {
+      try {
+        await permissionService.createRole("content_editor", [
+          {
+            method: "POST",
+            endpoint: "/products",
+          },
+          {
+            // Should fail since this is not a valid http request
+            method: "FETCH",
+            endpoint: "/products",
+          },
+        ])
+      } catch (err) {
+        expect(err.message).toEqual("Permission is not valid")
+      }
+    })
+
+    it("throws if role with name already exists", async () => {
+      try {
+        await permissionService.createRole("product_editor", [
+          {
+            method: "POST",
+            endpoint: "/order",
+          },
+        ])
+      } catch (err) {
+        expect(err.message).toEqual("product_editor already exists")
+      }
+    })
+  })
+
+  describe("grantRole", () => {
```

Comment:
Please add a test to grant a role to a user that doesn't yet have any roles: i.e. the roles field in metadata doesn't exist.